### PR TITLE
ES|QL: Fix possible ClassCastException with ReplaceMissingFieldWithNull

### LINKED ISSUE #110150: ES|QL: ReplaceMissingFieldWithNull could lead to ClassCastException
[ReplaceMissingFieldWithNull](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizer.java#L170) local logical optimization rule replaces expressions with a NULL `Literal` when a field is not mapped on local node.

This is dangerous because a `Literal` could not match the type of the expressions that are being replaced (eg. if the command expects a `NamedExpression` like in [this case](https://github.com/elastic/elasticsearch/issues/109974)), and it can trigger a ClassCastException.

With current feature set this optimization is safe, but we could miss the problem when we add new commands.

Notice that this optimization triggers a few other downstream optimizations that can be quite important, see `LocalLogicalPlanOptimizerTests.testMissingField*`

### PR DESCRIPTION
**Pull request by luigidellaquila**
Fix ReplaceMissingFieldWithNull by explicitly listing the commands that can be optimized replacing missing FieldAttributed with NULL Literals.

Also adding a unit test that demonstrates possible scenarios where introducing a new command can lead to `ClassCastException` with `ReplaceMissingFieldWithNull` local optimization rule and an integration test that covers https://github.com/elastic/elasticsearch/issues/109974

Fixes #110150

### CODE REVIEW

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Review by alex-spies — APPROVED** [review]
LGTM! Thanks Luigi!

**Review by astefan — COMMENTED** [review]
Thanks for adding the test, but what I think it would help us more understand these edge cases is an IT test.
From my previous attempts at reproducing this (single node, multiple indices, multiple queries in various forms, multiple shards - empty shards and shards with docs) I couldn't repro it. Multi-node, multi-cluster could be the deciding factor, not sure, but I think it would be a learning opportunity for us to see this edge case. Maybe we can improve further the code after finding the edge case.

**Comment by alex-spies** [comment]
> Thanks for adding the test, but what I think it would help us more understand these edge cases is an IT test. From my previous attempts at reproducing this (single node, multiple indices, multiple queries in various forms, multiple shards - empty shards and shards with docs) I couldn't repro it. Multi-node, multi-cluster could be the deciding factor, not sure, but I think it would be a learning opportunity for us to see this edge case. Maybe we can improve further the code after finding the edge case.

I don't think an IT makes sense, because for existing logical plan nodes `ReplaceMissingFieldWithNull` does the right thing (after Luigi fixed it). The problem is that it can silently break again if we change our logical plan nodes or create new ones.

The easiest would be to just change the default behavior and not replace missing field attributes by literal `NULL`s without knowing which logical plan node we apply this to.

My preference is that we do that change in this PR; that should be a very small local change and we can immediately unmute the test added by this.

**Comment by astefan** [comment]
@alex-spies noted.

I still want to understand when we encounter such a scenario, since none of us was able to reproduce it.

**Comment by luigidellaquila** [comment]
> I still want to understand when we encounter such a scenario, since none of us was able to reproduce it.

Actually the problem reproduces consistently in a specific multi-node environment.
The key is multi-node indeed: when a field is not present in any of the mappings of a data node, the optimization removes it locally and replaces it with a `Literal(null)`, triggering the `ClassCastException`.

I added an integration test that reproduces this scenario. The test is green, but if you revert [this fix](https://github.com/elastic/elasticsearch/pull/110096) you'll see it fail in multi-node.

> My preference is that we do that change in this PR; that should be a very small local change and we can immediately unmute the test added by this.

@alex-spies I agree with you, the fix is pretty straight-forward, I'll add it to this PR

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/plan/logical/Aggregate.java [hunk 1]
```diff
@@ -1,79 +0,0 @@
...
-import org.elasticsearch.xpack.esql.core.expression.Expression;
-import org.elasticsearch.xpack.esql.core.expression.Expressions;
-import org.elasticsearch.xpack.esql.core.expression.NamedExpression;
-import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
-import org.elasticsearch.xpack.esql.core.tree.Source;
-
-import java.util.List;
-import java.util.Objects;
-
-public class Aggregate extends UnaryPlan {
```
Comment (luigidellaquila) [inline x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/plan/logical/Aggregate.java:20]: Not used anymore, we have another `Aggregate` class in `esql` package

**Comment by luigidellaquila** [comment]
Done

**Review by alex-spies — APPROVED** [review]
Thanks Luigi!

**Comment by luigidellaquila** [comment]
Changed the title and description of the PR accordingly

**Review by astefan — APPROVED** [review]
LGTM. Thank you, @luigidellaquila

**Pull request merged**